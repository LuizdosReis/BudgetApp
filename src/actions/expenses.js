import uuid from 'uuid';
import database from '../firebase/firebase';

export const add = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSet = () => dispatch => database
  .ref('expenses')
  .once('value')
  .then((snapshot) => {
    const expenses = [];

    snapshot.forEach((childSnapshot) => {
      expenses.push({
        id: childSnapshot.key,
        ...childSnapshot.val(),
      });
    });

    dispatch(setExpenses(expenses));
  });

export const startAdd = (expenseData = {}) => (dispatch) => {
  const {
    description = '', category = null, createAt = 0, amount = 0,
  } = expenseData;

  const expense = {
    description,
    createAt,
    amount,
  };

  return database
    .ref('expenses')
    .push(expense)
    .then((ref) => {
      dispatch(
        add({
          id: ref.key,
          ...expense,
        }),
      );
    });
};

export const remove = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const edit = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});
