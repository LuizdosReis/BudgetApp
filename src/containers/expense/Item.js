import React from 'react';
import PropTypes from 'prop-types';

import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

const Item = ({
  amount, description, createAt, id,
}) => (
  <tr>
    <td className="table__element">
      <Link to={`/expenses/edit/${id}`}>{description}</Link>
    </td>
    <td>{numeral(amount / 100).format('$0,0.00')}</td>
    <td>{moment(createAt).format('DD/MM/YYYY')}</td>
  </tr>
);

Item.propTypes = {
  amount: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  createAt: PropTypes.number.isRequired,
  id: PropTypes.string.isRequired,
};

export default Item;
