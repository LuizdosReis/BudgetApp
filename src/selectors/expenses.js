import moment from 'moment';

export default (expenses, {
  startDate, endDate, description, sortBy,
}) => expenses
  .filter((expense) => {
    const createAtMoment = moment(expense.createAt);
    const startDateMatch = startDate ? createAtMoment.isSameOrAfter(startDate, 'day') : true;
    const endDateMatch = endDate ? createAtMoment.isSameOrBefore(endDate, 'day') : true;
    const descriptionMatch = description
      ? expense.description.toLowerCase().includes(description.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && descriptionMatch;
  })
  .sort((a, b) => {
    if (sortBy === 'description') {
      return a.description < b.description ? -1 : 1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? -1 : 1;
    }
    return a.createAt < b.createAt ? -1 : 1;
  });
