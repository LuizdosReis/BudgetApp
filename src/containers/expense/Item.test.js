import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Item from './Item';
import expenses from '../../fixtures/expenses';

test('should render Item with expense', () => {
  const wrapper = shallow(<Item {...expenses[0]} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
