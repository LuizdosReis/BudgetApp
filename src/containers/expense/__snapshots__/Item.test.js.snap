// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Item with expense 1`] = `
<tr>
  <td
    className="table__element"
  >
    <Link
      replace={false}
      to="/expenses/edit/0"
    >
      Gum
    </Link>
  </td>
  <td>
    $32.50
  </td>
  <td>
    01/01/1970
  </td>
</tr>
`;
